runtime/interpreter/value_storagecapabilitycontroller.go

```diff
@@ -19,115 +19,212 @@
-	capabilityID uint64,
-	borrowType StaticType,
-	delete func() error,
-	getTarget func() (PathValue, error),
-	retarget func(newPath PathValue) error,
+	memoryGauge common.MemoryGauge,
+	staticType StaticType,
+	targetPath PathValue,
+	capabilityID UInt64Value,
+) *StorageCapabilityControllerValue {
+	// Constant because its constituents are already metered.
+	common.UseMemory(memoryGauge, common.StorageCapabilityControllerValueMemoryUsage)
+	return NewUnmeteredStorageCapabilityControllerValue(
+		staticType,
+		targetPath,
+		capabilityID,
+	)
+}
+
+var _ Value = &StorageCapabilityControllerValue{}
+var _ atree.Value = &StorageCapabilityControllerValue{}
+var _ EquatableValue = &StorageCapabilityControllerValue{}
+var _ CapabilityControllerValue = &StorageCapabilityControllerValue{}
+var _ MemberAccessibleValue = &StorageCapabilityControllerValue{}
+
+func (*StorageCapabilityControllerValue) IsValue() {}
+
+func (*StorageCapabilityControllerValue) isCapabilityControllerValue() {}
+
+func (v *StorageCapabilityControllerValue) Accept(interpreter *Interpreter, visitor Visitor) {
+	visitor.VisitStorageCapabilityControllerValue(interpreter, v)
+}
+
+func (v *StorageCapabilityControllerValue) Walk(_ *Interpreter, walkChild func(Value)) {
+	walkChild(v.TargetPath)
+	walkChild(v.CapabilityID)
+}
+
+func (v *StorageCapabilityControllerValue) StaticType(_ *Interpreter) StaticType {
+	return PrimitiveStaticTypeStorageCapabilityController
+}
+
+func (*StorageCapabilityControllerValue) IsImportable(_ *Interpreter) bool {
+	return false
+}
+
+func (v *StorageCapabilityControllerValue) String() string {
+	return v.RecursiveString(SeenReferences{})
+}
+
+func (v *StorageCapabilityControllerValue) RecursiveString(seenReferences SeenReferences) string {
+	return format.StorageCapabilityController(
+		v.BorrowType.String(),
+		v.TargetPath.RecursiveString(seenReferences),
+		v.CapabilityID.RecursiveString(seenReferences),
+	)
+}
+
+func (v *StorageCapabilityControllerValue) MeteredString(memoryGauge common.MemoryGauge, seenReferences SeenReferences) string {
+	common.UseMemory(memoryGauge, common.StorageCapabilityControllerValueStringMemoryUsage)
+
+	return format.StorageCapabilityController(
+		v.BorrowType.MeteredString(memoryGauge),
+		v.CapabilityID.MeteredString(memoryGauge, seenReferences),
+		v.TargetPath.MeteredString(memoryGauge, seenReferences),
+	)
+}
+
+func (v *StorageCapabilityControllerValue) ConformsToStaticType(
+	_ *Interpreter,
+	_ LocationRange,
+	_ TypeConformanceResults,
+) bool {
+	return true
+}
+
+func (v *StorageCapabilityControllerValue) Equal(interpreter *Interpreter, locationRange LocationRange, other Value) bool {
+	otherController, ok := other.(*StorageCapabilityControllerValue)
+	if !ok {
+		return false
+	}
+
+	return otherController.TargetPath.Equal(interpreter, locationRange, v.TargetPath) &&
+		otherController.BorrowType.Equal(v.BorrowType) &&
+		otherController.CapabilityID.Equal(interpreter, locationRange, v.CapabilityID)
```

Comment:
Is it not necessary to compare the `RetargetFunction`?

Reply:
So far, everywhere else we only compare inner values, not functions. Functions are also not comparable.

Here, once could even argue that two controllers are equal if there ID is equal.